Add byteCast for conversions between different byte-sized types

darinadler opened the pull request:
#### dbd9819325e88d1d46f1662949df2edd3bba2666
<pre>
Add byteCast for conversions between different byte-sized types
<a href="https://bugs.webkit.org/show_bug.cgi?id=273380">https://bugs.webkit.org/show_bug.cgi?id=273380</a>
<a href="https://rdar.apple.com/127209334">rdar://127209334</a>

Reviewed by Chris Dumez.

* Source/JavaScriptCore/API/JSStringRef.cpp:
(JSStringGetUTF8CString): Use byteCast.

* Source/JavaScriptCore/assembler/MacroAssemblerPrinter.cpp:
(JSC::Printer::printMemory): Use static_cast instead of reinterpret_cast.
* Source/JavaScriptCore/assembler/ProbeFrame.h:
(JSC::Probe::Frame::Frame): Ditto.
* Source/JavaScriptCore/bytecode/UnlinkedMetadataTableInlines.h:
(JSC::UnlinkedMetadataTable::link): Ditto.

* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::Identifier::equal): Use byteCast.
(JSC::operator==): Ditto.
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::compareStrings const): Ditto.
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue): Ditto.

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::toStringWithRadixInternal): Removed unneeded static_cast.

* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::Segment::byte): Removed unneeded reinterpret_cast.

* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::fill): Use static_cast instead of reinterpret_cast.
(JSC::Wasm::Memory::copy): Ditto.
(JSC::Wasm::Memory::init): Ditto.

* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::dumpUChar32): Tweaked formatting.

* Source/WTF/wtf/FastFloat.cpp:
(WTF::parseDouble): Use byteCast.
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::toStdFileSystemPath): Ditto.
(WTF::FileSystemImpl::fromStdFileSystemPath): Ditto.

* Source/WTF/wtf/PageBlock.h:
(WTF::PageBlock::end const): Use static_cast instead of reinterpret_cast.

* Source/WTF/wtf/StdLibExtras.h: Added byteCast.

* Source/WTF/wtf/Threading.cpp:
(WTF::Thread::normalizeThreadName): Use byteCast.

* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL): Use span instead of dataAsUInt8Ptr.
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const): Ditto.

* Source/WTF/wtf/cf/VectorCF.h:
(WTF::span): Use static_cast instead of reinterpret_cast.

* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder&lt;CString&gt;::encodeForPersistence): Use span
instead of dataAsUInt8Ptr.
(WTF::Persistence::Coder&lt;CString&gt;::decodeForPersistence): Use byteCast.

* Source/WTF/wtf/text/AtomStringImpl.h: Use byteCast.
* Source/WTF/wtf/text/CString.h: Ditto.
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append): Ditto.
* Source/WTF/wtf/text/StringCommon.h: Ditto.
* Source/WTF/wtf/text/StringConcatenate.h: Ditto.
* Source/WTF/wtf/text/StringConcatenateNumbers.h: Ditto.
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equalInner): Ditto.
* Source/WTF/wtf/text/StringImpl.h: Ditto.
* Source/WTF/wtf/text/StringView.h: Ditto.
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::String): Ditto.
(WTF::String::fromUTF8): Ditto.
* Source/WTF/wtf/text/WTFString.h: Ditto.
* Source/WTF/wtf/text/cf/AtomStringImplCF.cpp:
(WTF::AtomStringImpl::add): Ditto.

* Source/WTF/wtf/unicode/Collator.h: Use char8_t instead of UTF8 in the
name for the collate function that takes null-terminated strings.
* Source/WTF/wtf/unicode/CollatorDefault.cpp:
(WTF::Collator::collate const): Use byteCast.
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::createIterator): Use byteCast.
(WTF::Collator::collate const): Use createIterator.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromVideoFrame): Use static_cast instead of reinterpret_cast.

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData): Use byteCast.
* Source/WebCore/PAL/pal/text/EncodingTables.cpp:
(PAL::jis0208): Ditto.
(PAL::jis0212): Ditto.
(PAL::big5): Ditto.
(PAL::eucKR): Ditto.
(PAL::gb18030): Ditto.
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::TextCodecCJK::eucJPDecode): Ditto.
(PAL::TextCodecCJK::big5Decode): Ditto.
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::decode): Ditto.
(PAL::TextCodecICU::encode const): Ditto.
* Source/WebCore/PAL/pal/text/TextEncodingDetectorICU.cpp:
(PAL::detectTextEncoding): Ditto.
* Source/WebCore/css/process-css-pseudo-selectors.py: Ditto.
* Source/WebCore/html/track/VTTScanner.h:
(WebCore::VTTScanner::scan): Ditto.

* Source/WebCore/html/track/WebVTTTokenizer.cpp:
(WebCore::equalLiteral): Deleted.

* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::showText): Use span instead of dataAsUInt8Ptr.

* Source/WebCore/platform/SharedBuffer.h: Deleted dataAsCharPtr.

* Source/WebCore/platform/SharedBufferChunkReader.cpp:
(WebCore::SharedBufferChunkReader::nextChunk): Use byteCast.
* Source/WebCore/platform/graphics/MIMESniffer.cpp:
(WebCore::MIMESniffer::span8): Ditto.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::initialize): Ditto.
(WebCore::GraphicsContextGLANGLE::getString): Ditto.
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels): Ditto.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::defaultVariationValues): Use assignment instead of static_cast.

* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample): Use byteCast.
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData): Ditto.
* Source/WebCore/platform/graphics/cpu/arm/filters/FEBlendNeonApplier.cpp:
(WebCore::FEBlendNeonApplier::applyPlatform const): Ditto.
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(WebCore::ScalableImageDecoder::create): Ditto.

* Source/WebCore/platform/network/cf/ResourceRequestCFNet.h:
(WebCore::httpHeaderValueUsingSuitableEncoding): Use span instead of
dataAsUInt8Ptr.
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::platformCertificateInfo const): Ditto.

* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::toString const): Use byteCast.
(WebCore::ScriptBuffer::append): Ditto. Also fixed a mistake where this was
using uint8_t instead of char8_t.

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc): Use byteCast.
(WebCore::writeToStringBuilder): Ditto.
(WebCore::XSLTProcessor::transformToString): Ditto.
* Source/WebCore/xml/XSLTUnicodeSort.cpp:
(WebCore::xsltUnicodeSortFunction): Ditto.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLParserContext::createMemoryParser): Use byteCast.
(WebCore::toString): Ditto.
(WebCore::toAtomString): Ditto.
(WebCore::sharedXHTMLEntity): Ditto.
(WebCore::convertUTF16EntityToUTF8): Ditto.
(WebCore::getXHTMLEntity): Ditto.
(WebCore::xmlDocPtrForString): Ditto.
(WebCore::attributesStartElementNsHandler): Ditto.
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const): Ditto.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(getBufferSubData): Ditto.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent): Ditto.
* Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:
(WebKit::LibWebRTCSocketClient::signalReadPacket): Ditto.
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl): Ditto.

* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCopyCFURL): Use span instead of dataAsUInt8Ptr.
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl): Ditto.

* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::operator&lt;&lt;): Use byteCast.
(WebKit::encodeLegacySessionState): Use checked_cf_cast.

* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::operator&lt;&lt;): Use byteCast.
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.cpp:
(WebKit::LibWebRTCSocket::signalReadPacket): Ditto.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityTreeData): Use checked_cf_cast.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook): Use byteCast.
* Source/WebKitLegacy/mac/Misc/WebNSDataExtras.mm:
(-[NSData _webkit_guessedMIMETypeForXML]): Ditto.

* Tools/TestWebKitAPI/Tests/WTF/StdLibExtrasTests.cpp:
(TestWebKitAPI::TEST(WTF_StdLibExtras, ByteCast)): Added some tests.

* Tools/TestWebKitAPI/Tests/WTF/StringBuilder.cpp: Rewrite tests so it&apos;s easier
to see which one failed. Did not use byteCast here.

* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST(WTF, StringImplEqualIgnoringASCIICaseBasic)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEqualIgnoringASCIICaseWithEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplFindIgnoringASCIICaseOnEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplStartsWithIgnoringASCIICaseWithNull)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplStartsWithIgnoringASCIICaseWithEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEndsWithIgnoringASCIICaseWithNull)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEndsWithIgnoringASCIICaseWithEmpty)): Ditto.
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST(WTF, StringViewEqualIgnoringASCIICaseBasic)): Ditto.
(TestWebKitAPI::TEST(WTF, StringViewEqualIgnoringASCIICaseWithEmpty)): Ditto.

* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::eq): Remomved unneeded reinterpret_cast.

* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
(TestWebKitAPI::TEST_F(FragmentedSharedBufferTest, getSomeData)): Use byteCast
and span instead of dataAsCharPtr.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/PDFLinkReferrer.mm:
(TEST(WebKit, PDFLinkReferrer)): Use byteCast.

* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.cpp:
(TestWebKitAPI::CGImagePixelReader::at const): Use static_cast instead of
reinterpret_cast.

* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::proxyDefinition): Use byteCast.
(TestWebKitAPI::Connection::webSocketHandshake): Ditto.

Canonical link: <a href="https://commits.webkit.org/279338@main">https://commits.webkit.org/279338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c704904e024c47d73332159b5934541dab5fad31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5688 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56479 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55505 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39791 "Failed to compile WebKit") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43128 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2550 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55298 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24258 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3258 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2082 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46556 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58075 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52713 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28343 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3378 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50530 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29561 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46153 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49846 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30481 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65018 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7819 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29317 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12372 "Found 6 new JSC stress test failures: stress/spread-non-array.js.no-llint, wasm.yaml/wasm/function-tests/add-12.js.wasm-eager, wasm.yaml/wasm/function-tests/br-if-as-return.js.wasm-eager, wasm.yaml/wasm/function-tests/brTableAsIf.js.wasm-eager, wasm.yaml/wasm/function-tests/dead-call.js.wasm-eager, wasm.yaml/wasm/regress/242294.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -719,6 +719,39 @@ void memsetSpan(std::span<T, Extent> destination, uint8_t byte)
+
+template<typename> struct ByteCastTraits;
+
+template<ByteType T> struct ByteCastTraits<T> {
```
Is the style we're using for concepts to put them as template parameters?

I've been using `requires (myConcept)` in https://github.com/WebKit/WebKit/pull/27753 but it feels like we should come up with a consistent style. I went with `requires` because it's an easier drop in replacement for `enable_if` and IIUC you can have compound constraints without having to create a name for it.

darinadler replied:
I used the style that I prefer for this code. I would have used `requires` if it was a more complex case that required it. while I absolutely think it will be great for WebKit to have consistent style, and I strongly oppose the idea of different "JavaScriptCore style', for example, it seems unnecessary to have that style require the same syntax for both simple and more complex cases. I don't think adding both the `typename` and `requires` keywords would make these templates better. Similarly, I think we can and should use `auto` instead of writing out templates the old way in new code, despite the fact that we haven't done it before.

darinadler replied:
Seems a little premature to codify a style rule for this, but as soon as the preferred style emerges from our work, we can write it down.

kmiller68 replied:
Yeah, I'm not proposing a different "JavaScriptCore style".

darinadler committed:
Add byteCast for conversions between different byte-sized types
https://bugs.webkit.org/show_bug.cgi?id=273380
rdar://127209334

Reviewed by Chris Dumez.

* Source/JavaScriptCore/API/JSStringRef.cpp:
(JSStringGetUTF8CString): Use byteCast.

* Source/JavaScriptCore/assembler/MacroAssemblerPrinter.cpp:
(JSC::Printer::printMemory): Use static_cast instead of reinterpret_cast.
* Source/JavaScriptCore/assembler/ProbeFrame.h:
(JSC::Probe::Frame::Frame): Ditto.
* Source/JavaScriptCore/bytecode/UnlinkedMetadataTableInlines.h:
(JSC::UnlinkedMetadataTable::link): Ditto.

* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::Identifier::equal): Use byteCast.
(JSC::operator==): Ditto.
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::compareStrings const): Ditto.
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue): Ditto.

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::toStringWithRadixInternal): Removed unneeded static_cast.

* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::Segment::byte): Removed unneeded reinterpret_cast.

* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::fill): Use static_cast instead of reinterpret_cast.
(JSC::Wasm::Memory::copy): Ditto.
(JSC::Wasm::Memory::init): Ditto.

* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::dumpUChar32): Tweaked formatting.

* Source/WTF/wtf/FastFloat.cpp:
(WTF::parseDouble): Use byteCast.
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::toStdFileSystemPath): Ditto.
(WTF::FileSystemImpl::fromStdFileSystemPath): Ditto.

* Source/WTF/wtf/PageBlock.h:
(WTF::PageBlock::end const): Use static_cast instead of reinterpret_cast.

* Source/WTF/wtf/StdLibExtras.h: Added byteCast.

* Source/WTF/wtf/Threading.cpp:
(WTF::Thread::normalizeThreadName): Use byteCast.

* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::userVisibleURL): Use span instead of dataAsUInt8Ptr.
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const): Ditto.

* Source/WTF/wtf/cf/VectorCF.h:
(WTF::span): Use static_cast instead of reinterpret_cast.

* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder<CString>::encodeForPersistence): Use span
instead of dataAsUInt8Ptr.
(WTF::Persistence::Coder<CString>::decodeForPersistence): Use byteCast.

* Source/WTF/wtf/text/AtomStringImpl.h: Use byteCast.
* Source/WTF/wtf/text/CString.h: Ditto.
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::append): Ditto.
* Source/WTF/wtf/text/StringCommon.h: Ditto.
* Source/WTF/wtf/text/StringConcatenate.h: Ditto.
* Source/WTF/wtf/text/StringConcatenateNumbers.h: Ditto.
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equalInner): Ditto.
* Source/WTF/wtf/text/StringImpl.h: Ditto.
* Source/WTF/wtf/text/StringView.h: Ditto.
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::String): Ditto.
(WTF::String::fromUTF8): Ditto.
* Source/WTF/wtf/text/WTFString.h: Ditto.
* Source/WTF/wtf/text/cf/AtomStringImplCF.cpp:
(WTF::AtomStringImpl::add): Ditto.

* Source/WTF/wtf/unicode/Collator.h: Use char8_t instead of UTF8 in the
name for the collate function that takes null-terminated strings.
* Source/WTF/wtf/unicode/CollatorDefault.cpp:
(WTF::Collator::collate const): Use byteCast.
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::createIterator): Use byteCast.
(WTF::Collator::collate const): Use createIterator.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromVideoFrame): Use static_cast instead of reinterpret_cast.

* Source/WebCore/Modules/encryptedmedia/InitDataRegistry.cpp:
(WebCore::isPlayReadySanitizedInitializationData): Use byteCast.
* Source/WebCore/PAL/pal/text/EncodingTables.cpp:
(PAL::jis0208): Ditto.
(PAL::jis0212): Ditto.
(PAL::big5): Ditto.
(PAL::eucKR): Ditto.
(PAL::gb18030): Ditto.
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::TextCodecCJK::eucJPDecode): Ditto.
(PAL::TextCodecCJK::big5Decode): Ditto.
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::decode): Ditto.
(PAL::TextCodecICU::encode const): Ditto.
* Source/WebCore/PAL/pal/text/TextEncodingDetectorICU.cpp:
(PAL::detectTextEncoding): Ditto.
* Source/WebCore/css/process-css-pseudo-selectors.py: Ditto.
* Source/WebCore/html/track/VTTScanner.h:
(WebCore::VTTScanner::scan): Ditto.

* Source/WebCore/html/track/WebVTTTokenizer.cpp:
(WebCore::equalLiteral): Deleted.

* Source/WebCore/page/cocoa/ResourceUsageOverlayCocoa.mm:
(WebCore::showText): Use span instead of dataAsUInt8Ptr.

* Source/WebCore/platform/SharedBuffer.h: Deleted dataAsCharPtr.

* Source/WebCore/platform/SharedBufferChunkReader.cpp:
(WebCore::SharedBufferChunkReader::nextChunk): Use byteCast.
* Source/WebCore/platform/graphics/MIMESniffer.cpp:
(WebCore::MIMESniffer::span8): Ditto.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::initialize): Ditto.
(WebCore::GraphicsContextGLANGLE::getString): Ditto.
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels): Ditto.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::defaultVariationValues): Use assignment instead of static_cast.

* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample): Use byteCast.
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::parseOpusPrivateData): Ditto.
* Source/WebCore/platform/graphics/cpu/arm/filters/FEBlendNeonApplier.cpp:
(WebCore::FEBlendNeonApplier::applyPlatform const): Ditto.
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(WebCore::ScalableImageDecoder::create): Ditto.

* Source/WebCore/platform/network/cf/ResourceRequestCFNet.h:
(WebCore::httpHeaderValueUsingSuitableEncoding): Use span instead of
dataAsUInt8Ptr.
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::platformCertificateInfo const): Ditto.

* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::toString const): Use byteCast.
(WebCore::ScriptBuffer::append): Ditto. Also fixed a mistake where this was
using uint8_t instead of char8_t.

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::docLoaderFunc): Use byteCast.
(WebCore::writeToStringBuilder): Ditto.
(WebCore::XSLTProcessor::transformToString): Ditto.
* Source/WebCore/xml/XSLTUnicodeSort.cpp:
(WebCore::xsltUnicodeSortFunction): Ditto.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLParserContext::createMemoryParser): Use byteCast.
(WebCore::toString): Ditto.
(WebCore::toAtomString): Ditto.
(WebCore::sharedXHTMLEntity): Ditto.
(WebCore::convertUTF16EntityToUTF8): Ditto.
(WebCore::getXHTMLEntity): Ditto.
(WebCore::xmlDocPtrForString): Ditto.
(WebCore::attributesStartElementNsHandler): Ditto.
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const): Ditto.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(getBufferSubData): Ditto.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent): Ditto.
* Source/WebKit/NetworkProcess/webrtc/LibWebRTCSocketClient.cpp:
(WebKit::LibWebRTCSocketClient::signalReadPacket): Ditto.
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl): Ditto.

* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCopyCFURL): Use span instead of dataAsUInt8Ptr.
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl): Ditto.

* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::operator<<): Use byteCast.
(WebKit::encodeLegacySessionState): Use checked_cf_cast.

* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
(WebKit::HistoryEntryDataEncoder::operator<<): Use byteCast.
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.cpp:
(WebKit::LibWebRTCSocket::signalReadPacket): Ditto.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getAccessibilityTreeData): Use checked_cf_cast.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook): Use byteCast.
* Source/WebKitLegacy/mac/Misc/WebNSDataExtras.mm:
(-[NSData _webkit_guessedMIMETypeForXML]): Ditto.

* Tools/TestWebKitAPI/Tests/WTF/StdLibExtrasTests.cpp:
(TestWebKitAPI::TEST(WTF_StdLibExtras, ByteCast)): Added some tests.

* Tools/TestWebKitAPI/Tests/WTF/StringBuilder.cpp: Rewrite tests so it's easier
to see which one failed. Did not use byteCast here.

* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST(WTF, StringImplEqualIgnoringASCIICaseBasic)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEqualIgnoringASCIICaseWithEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplFindIgnoringASCIICaseOnEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplStartsWithIgnoringASCIICaseWithNull)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplStartsWithIgnoringASCIICaseWithEmpty)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEndsWithIgnoringASCIICaseWithNull)): Ditto.
(TestWebKitAPI::TEST(WTF, StringImplEndsWithIgnoringASCIICaseWithEmpty)): Ditto.
* Tools/TestWebKitAPI/Tests/WTF/StringView.cpp:
(TestWebKitAPI::TEST(WTF, StringViewEqualIgnoringASCIICaseBasic)): Ditto.
(TestWebKitAPI::TEST(WTF, StringViewEqualIgnoringASCIICaseWithEmpty)): Ditto.

* Tools/TestWebKitAPI/Tests/WebCore/CBORWriterTest.cpp:
(TestWebKitAPI::eq): Remomved unneeded reinterpret_cast.

* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
(TestWebKitAPI::TEST_F(FragmentedSharedBufferTest, getSomeData)): Use byteCast
and span instead of dataAsCharPtr.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/PDFLinkReferrer.mm:
(TEST(WebKit, PDFLinkReferrer)): Use byteCast.

* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.cpp:
(TestWebKitAPI::CGImagePixelReader::at const): Use static_cast instead of
reinterpret_cast.

* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::proxyDefinition): Use byteCast.
(TestWebKitAPI::Connection::webSocketHandshake): Ditto.

Canonical link: https://commits.webkit.org/279338@main
